homeassistant/components/google_assistant/trait.py

```diff
@@ -609,6 +614,102 @@ async def execute(self, command, data, params, challenge):
+        elif 75 <= battery_level < 100:
+            descriptive_capacity_remaining = "HIGH"
+        elif 50 <= battery_level < 75:
+            descriptive_capacity_remaining = "MEDIUM"
+        elif 25 <= battery_level < 50:
+            descriptive_capacity_remaining = "LOW"
+        elif 0 <= battery_level < 25:
+            descriptive_capacity_remaining = "CRITICALLY_LOW"
+        return {
+            "descriptiveCapacityRemaining": descriptive_capacity_remaining,
+            "capacityRemaining": [{"rawValue": battery_level, "unit": "PERCENTAGE"}],
+            "capacityUntilFull": [
+                {"rawValue": 100 - battery_level, "unit": "PERCENTAGE"}
+            ],
+            "isCharging": self.state.state == vacuum.STATE_DOCKED,
+            "isPluggedIn": self.state.state == vacuum.STATE_DOCKED,
+        }
+
+    async def execute(self, command, data, params, challenge):
+        """Execute a dock command to start charging."""
+        if params["charge"]:
+            if self.state.state == vacuum.STATE_DOCKED:
+                raise SmartHomeError(
+                    ERR_ALREADY_DOCKED,
+                    "The vacuum is already docked",
+                )
+            else:
+                await self.hass.services.async_call(
+                    self.state.domain,
+                    vacuum.SERVICE_RETURN_TO_BASE,
+                    {ATTR_ENTITY_ID: self.state.entity_id},
+                    blocking=True,
+                    context=data.context,
+                )
+        else:
+            if self.state.state == vacuum.STATE_DOCKED:
+                if (
+                    self.state.attributes.get(ATTR_SUPPORTED_FEATURES, 0)
+                    & vacuum.SUPPORT_PAUSE
+                ):
+                    await self.hass.services.async_call(
+                        self.state.domain,
+                        vacuum.SERVICE_START_PAUSE,
+                        {ATTR_ENTITY_ID: self.state.entity_id},
+                        blocking=True,
+                        context=data.context,
+                    )
+                elif (
+                    self.state.attributes.get(ATTR_SUPPORTED_FEATURES, 0)
+                    & vacuum.SUPPORT_START
+                ):
+                    await self.hass.services.async_call(
+                        self.state.domain,
+                        vacuum.SERVICE_START,
+                        {ATTR_ENTITY_ID: self.state.entity_id},
+                        blocking=True,
+                        context=data.context,
+                    )
+                else:
+                    raise SmartHomeError(
+                        ERR_FUNCTION_NOT_SUPPORTED,
+                        "Cannot stop charging a docked vacuum",
+                    )
+            else:
+                raise SmartHomeError(
+                    ERR_UNAVAILABLE_WHILE_RUNNING,
+                    "Cannot stop charging an undocked vacuum",
+                )
```

Comment:
We prefer early returns. So please do like:
```
if docked (or not):
  raise ...

# rest of code without else.
```

In both the charge and the non charge case.

Reply:
Fine for the "charge" case, but in the "non charge" case the late return is a fallback case, an early one will make the if statements unnecessarily complicated

Reply:
now irrelevant